File: call_thing.rb
#!/usr/bin/env ruby

"use strict";

class Thing
  attr_accessor :_actions
  def initialize
    self._actions = {};
  end
end

class Thing
  def add(name, funct)
    if self._actions.has_key?(name)
      self._actions[name] << funct
    else
      self._actions[name] = [funct]
    end
  end

  def call(name, value)
    0.upto(self._actions[name].size - 1) do |i|
      return self._actions[name][i].call(value)
    end
  end
end

x = Thing.new
x.add 'thing_1', lambda {|x| x}
x.add 'thing_1', lambda {|y| y + y}
x.add 'thing_2', lambda {|z| z.to_i - 1}
puts x.call 'thing_1', 'ooo'
puts x.call 'thing_1', 2
puts x.call 'thing_2', 'ooo'
puts x.call 'thing_2', 2

puts x.inspect

__END__

ooo
2
-1
1
#<Thing:0x007f80a40ba168 @_actions={
  "thing_1"=>[#<Proc:0x007f80a40ba0f0@./call_event.rb:29 (lambda)>,
  #<Proc:0x007f80a40ba050@./call_event.rb:30 (lambda)>],
  "thing_2"=>[#<Proc:0x007f80a40ba000@./call_event.rb:31 (lambda)>]}>

File: call_thing.js
#!/usr/bin/env node

"use strict";

var Thing = function() {
  this._actions = {};
}

Thing.prototype.add = function(name, funct) {
  if(this._actions[name]) {
    this._actions[name].push(funct);
  } else {
    this._actions[name] = [funct];
  }
}

Thing.prototype.call = function(name, value) {
  for (var i = 0, f; (f = this._actions[name][i]); i++) {
    f(value);
  }
}

var x = new Thing;
x.add('thing_1', function(x){console.log(x)});
x.add('thing_1', function(y){console.log(y + y)});
x.add('thing_2', function(z){console.log(z - 1)});
x.call('thing_1', 'ooo');
x.call('thing_1', 2);
x.call('thing_2', 'ooo');
x.call('thing_2', 2);

console.log(x);

/*

ooo 
oooooo 
2 
4 
NaN
1 
Object { 
  _actions: Object {
  thing_1: [ [Function], [Function] ],
  thing_2: [ [Function] ] } } 

*/
